src/utils.ts

```diff
@@ -192,3 +191,52 @@ export class IndexBytes {
     }
   }
 }
+
+/**
+ * The hash length has to be 31 instead of 32 that comes from the keccak hash function
+ */
+export function serializeVersion(version: MarshalVersion): Bytes<31> {
+  const versionName = 'mantaray'
+  const versionSeparator = ':'
+  const hashBytes = keccak256Hash(versionName + versionSeparator + version)
+
+  return hashBytes.slice(0, 31) as Bytes<31>
+}
+
+export function serializeMedata(metadata: MetadataMapping): Uint8Array {
+  const jsonString = JSON.stringify(metadata)
+
+  return new TextEncoder().encode(jsonString)
+}
+
+/** Returns segment padded stringified JSON byte array */
+export function serializeMetadataInSegment(metadata: MetadataMapping | undefined, segmentSize: number): Uint8Array {
+  if (!metadata) {
+    const bytes = new Uint8Array(segmentSize * 32)
+    bytes.fill(32) // space padding
+
+    return bytes
+  }
+
+  const jsonString = JSON.stringify(metadata)
+  const jsonData = new TextEncoder().encode(jsonString)
+  const remainingSegmentBytes = segmentSize * 32 - jsonData.length
+
+  if (remainingSegmentBytes < 0) throw new Error('segmentSize for fork metadata is smaller than the jsonData')
+
+  const paddingBytes = new Uint8Array(remainingSegmentBytes)
+  paddingBytes.fill(32) // space padding
+
+  return new Uint8Array([...jsonData, ...paddingBytes])
+}
+
+/** If the JSON deserialisation of the data is not succesful, it will give back undefined  */
+export function deserializeMetadata(data: Uint8Array): MetadataMapping | undefined {
```

Comment:
no compatibility issues here, just the bytesize always reserved for the JSON metadata that causes issue if there wasn't any metadata defined for the fork, because the zero bytearray not a valid stringified JSON object.